Refactored HomePage route handler to fetch initClassInfo and getFacil…

shubh1007 opened the pull request:
### Summary

This is the continuation of the refactoring process for the coach page, transitioning from a global route handler to local route handlers. #11219

### References

#### List of all planRoutes:

| Route name                  | Handler/component to refactor | Refactor done? | How to test the page                                       |
|-----------------------------|-------------------------------|----------------|-----------------------------------------------------------|
| PageNames.HOME_PAGE         | showHomePage      | yes              | Coach -> Class Home                                              |
| HomeActivityPage.name             | showHomePage                | yes            | Coach -> Class Home -> activity                                    |

### Reviewer guidance

Please follow the table for reviewing the refactored routes.

### Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

### Reviewer checklist

- [ ] Automated test coverage is satisfactory
- [ ] PR is fully functional
- [ ] PR has been tested for accessibility regressions
- [ ] Documentation is updated
- [ ] Contributor is in AUTHORS.md

shubh1007 committed:
Refactored HomePage route handler to fetch initClassInfo and getFacilities data individually

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

MisRob commented on kolibri/plugins/coach/assets/src/app.js:
```diff
@@ -84,6 +86,8 @@ class CoachToolsModule extends KolibriApp {
           GroupsPage.name,
           GroupMembersPage.name,
           GroupEnrollPage.name,
+          HomePage.name,
```
Since the related route uses `PageNames.HOME_PAGE` name, I think it'd be safer to use it here as well (it's possible they're same but as far as I can say that'd be rather a coincidence.

MisRob commented on kolibri/plugins/coach/assets/src/routes/index.js:
```diff
@@ -58,10 +70,11 @@ export default [
     path: '/:classId?/home',
     component: HomePage,
-    handler: (toRoute, fromRoute, next) => {
+    handler: async (toRoute, fromRoute, next) => {
```
I think that here `async` could be removed because in this function we're not `await`ing, or was there some trouble without it? Same on line 87.

shubh1007 replied:
Yes I tried this approach but it was not able to load the promises properly. So, I added async here.

MisRob replied:
I am not sure if I can understand what exactly you mean "it was not able to load the promises properly".

If I remove `async` from here, it doesn't throw any errors and works like before in the current code, because we're not calling `await showHomePage(toRoute)`. However, because of that missing `await`, I don't think that promises get resolved before `'notLoading'`.

Overall looking at the way you implemented it, I believe the intention actually was to wait for these two promises to be finished before calling `'notLoading'`, which we indeed need here. To achieve that, we could:

```
function showHomePage(toRoute) {
  const initClassInfoPromise = store.dispatch('initClassInfo', toRoute.params.classId);
  const getFacilitiesPromise =
    store.getters.isSuperuser && store.state.core.facilities.length === 0
      ? store.dispatch('getFacilities').catch(() => {})
      : Promise.resolve();

  return Promise.all([initClassInfoPromise, getFacilitiesPromise]);
}
```

and then

```
{
    name: PageNames.HOME_PAGE,
    path: '/:classId?/home',
    component: HomePage,
    handler: async (toRoute, fromRoute, next) => {
      if (classIdParamRequiredGuard(toRoute, HomePage.name, next)) {
        return;
      }
      await showHomePage(toRoute);
      store.dispatch('notLoading');
    },
```

Is this what you've meant to achieve?

MisRob replied:
I think this is the last thing to cleanup - seems to be caused by playing around with two possible placements of `store.dispatch('notLoading');` that got mixed up. Since now it's removed from the helper `showHomePage` function, we need to update the structure a bit.

MisRob commented on kolibri/plugins/coach/assets/src/routes/index.js:
```diff
@@ -13,6 +13,18 @@ import reportRoutes from './reportRoutes';
+      : Promise.resolve();
+
+  await Promise.all([initClassInfoPromise, getFacilitiesPromise]);
+  store.dispatch('notLoading');
```
`notLoading` is now called twice for each of the routes being refactored - I'd suggest to remove it from here, or from lines 78 and 89.

shubh1007 replied:
Resolved.

MisRob commented on kolibri/plugins/coach/assets/src/routes/index.js:
```diff
@@ -13,6 +13,18 @@ import reportRoutes from './reportRoutes';
 import { classIdParamRequiredGuard } from './utils';
 
+async function showHomePage(toRoute) {
+  console.log('Inside showHomePage');
```
Please remove `console.log`. I am not sure why lint check didn't detect it.

shubh1007 replied:
I have removed it.

MisRob commented:
Hi @shubh1007, this is great! I can see you grasped the problem and the strategy for solving it, thanks a lot. This will really help us to unblock some important work. Leaving just a few details, but overall looks really good.

Do you wish to continue with some other routes from this file? If so, would you push them to this PR as well or would you rather work in a new one?

Note that I'm mostly offline until July 8 - just chiming in on this issue to confirm that you can continue if you'd like, and on some days I will be available to answer questions.

MisRob commented:
@shubh1007 I don't think you're on Slack yet, but recently I approached the community of developers with this task and since it's pretty abstract, I also wrote something about what it will have impact on in the near future, if that'd be of your interest:

![Screenshot from 2024-06-25 20-49-22](https://github.com/learningequality/kolibri/assets/13509191/d20b710f-f3bd-4573-a384-762f79bd4eea)

And also it showed up that the upcoming 0.18 Kolibri version will benefit from these updates to unblock some major improvements to the navigation UX. So thanks a lot, despite this is not so visible at this point, coaches using Kolibri will notice it soon :) !

shubh1007 committed:
resolved the reviewed code.

shubh1007 committed:
resolved issue of naming the routes

shubh1007 committed:
Merge branch 'learningequality:develop' into HomeRoute

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

shubh1007 committed:
structured the showHomePage function

shubh1007 committed:
structured the showHomePage function

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

MisRob commented:
Thanks for following up @shubh1007. I will finish review when I'm back online next week. Meanwhile, I feel confident that we're on the same page about this task so feel free to continue working on other routes if you'd like (or please message us if you don't want to work on more so we know we should re-assign)

MisRob committed:
Fix colors of list items

and remove hardcoded colors
in favor of theme tokens.

Fixes https://github.com/learningequality/kolibri/issues/12425

MisRob committed:
Add key to v-for

This fixes vue/require-v-for-key 'Elements in iteration expect to have 'v-bind:key' directives.'
lint error and ensures efficient Vue updates.

MisRob committed:
Fix vue/no-unused-components

MisRob committed:
Disable linter for sections prop

MisRob committed:
Fixes kolibri/vue-no-unused-properties

MisRob committed:
Fixes kolibri/vue-no-unused-methods

MisRob committed:
Disable linter for selectedExercises prop

MisRob committed:
Merge branch 'develop' into HomeRoute

MisRob committed:
Fix missing await

MisRob committed:
Fix page name

There's no PageNames.HomeActivityPage

MisRob committed:
Cleanup rebase

MisRob reviewed (approved):
@shubh1007 Thank you. I pushed two last tweaks. See commit messages to understand. As soon as checks pass, we will merge.
